fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingBase.java

```diff
@@ -885,6 +954,56 @@ protected void validateSameMetadataOrThrow(FDBRecordStore store) {
         }
     }
 
+    CompletableFuture<Void> indexingStamp(@Nonnull ConcurrentHashMap<String, IndexBuildProto.IndexBuildIndexingStamp> oldStamps,
+                                          @Nullable OnlineIndexer.IndexingStampOperation op,
+                                          @Nullable String id,
+                                          @Nullable  Integer timeoutSeconds) {
+        return getRunner().runAsync(context -> openRecordStore(context).thenCompose(store ->
+            forEachTargetIndex(index -> store.loadIndexingTypeStampAsync(index)
+                    .thenApply(stamp -> indexingStamp(oldStamps, store, index, stamp, op, id, timeoutSeconds)))
+        ));
+    }
+
+    boolean indexingStamp(@Nonnull ConcurrentHashMap<String, IndexBuildProto.IndexBuildIndexingStamp> oldStamps,
+                          @Nonnull FDBRecordStore store,
+                          @Nonnull Index index,
+                          @Nullable IndexBuildProto.IndexBuildIndexingStamp stamp,
+                          @Nullable OnlineIndexer.IndexingStampOperation op,
+                          @Nullable String id,
+                          @Nullable  Integer timeoutSeconds
+    ) {
+        oldStamps.put(index.getName(), stamp != null ? stamp :
+                                       IndexBuildProto.IndexBuildIndexingStamp.newBuilder()
+                                               .setMethod(IndexBuildProto.IndexBuildIndexingStamp.Method.INVALID)
+                                               .build());
+        if (op == null || stamp == null || op.equals(OnlineIndexer.IndexingStampOperation.QUERY)) {
+            return false;
+        }
+
+        IndexBuildProto.IndexBuildIndexingStamp.Builder builder;
+        try {
+            builder = IndexBuildProto.IndexBuildIndexingStamp.parseFrom(stamp.toByteArray()).toBuilder();
+        } catch (InvalidProtocolBufferException e) {
+            return false;
+        }
+
+        if (op == OnlineIndexer.IndexingStampOperation.BLOCK) {
+            builder.setBlock(true);
+            if (id != null) {
+                builder.setBlockID(id);
+            }
+            if (timeoutSeconds != null && timeoutSeconds > 0) {
+                builder.setBlockExpireEpochSeconds((int)(System.currentTimeMillis() / 1000) + timeoutSeconds);
+            }
+        }
+        if (op == OnlineIndexer.IndexingStampOperation.UNBLOCK &&
+                (id == null || id.isEmpty() || id.equals(stamp.getBlockID()))) {
+            builder.setBlock(false);
+        }
+        store.saveIndexingTypeStamp(index, builder.build());
```

Comment:
Yes. 
It can be especially useful as a way to clear halt mutual indexing, then continue with fewer workers.

Reply:
Is the information there (whatever is used to "stop mutual indexing") dropped from the type stamp in the `BLOCK` or `UNBLOCK` case? I feel like you'd want to get the value of the type stamp from after the operation has applied on these non-read only cases, but I could be wrong.

There's also some weirdness here with transaction retry loops: if the first transaction updates the stamp, and then commits, then if it hits `commit_unknown_result`, it can see the new value on retry. So, the exact thing that this method returns is a little more unpredictable than you'd like.

Reply:
Are there retries in this path?

Reply:
Yes, because this is called within a lambda that gets passed to `getRunner().runAsync`, which will then retry the inner operation